pkg/monitors/elasticsearch/query/response.go

```diff
@@ -0,0 +1,174 @@
+package query
+
+import (
+	"encoding/json"
+)
+
+// Represents all kinds of supported aggregations. aggregationResponse type can
+// be determined in the following manner -
+// 1) single bucket aggregationInfo contains only "doc_count" and a map to sub aggregations
+// 2) multi bucket aggregationInfo contains "buckets"and a map to sub aggregations
+// 3) metric aggregationInfo contains "value" or "values"
+// These structs are defined based on inputs from
+// https://github.com/elastic/elasticsearch/tree/master/server/src/main/java/org/elasticsearch/search/aggregations
+type aggregationResponse struct {
+	// Number of documents backing the aggregationInfo
+	DocCount *int64 `json:"doc_count,omitempty"`
+	// Non nil for single value metric aggregations
+	Value interface{} `json:"value,omitempty"`
+	// Non nil for multi-value metrics aggregations
+	Values interface{} `json:"values,omitempty"`
+	// Map of sub aggregations with aggregationInfo names, includes both bucket
+	// and metric aggregations
+	SubAggregations map[string]aggregationResponse `json:"-"`
+	// Non nil for multi-value bucket aggregations
+	Buckets []*bucketResponse `json:"buckets,omitempty"`
+	// This field only exists to handle cases where buckets are returned as a
+	// map from the key to the object itself. In such cases the "buckets" field
+	// will be post processed and put into the "Buckets" field of this struct
+	BucketsMap map[string]*bucketResponse `json:"buckets,omitempty"`
```

Comment:
What exactly does the json module do when you have two separate struct fields with the same json key `buckets`?  In yaml this will cause a panic, but not sure how json handles it.

Reply:
With `json` from what I've seen depending on the type of `buckets`, it intelligently does the right thing. But I noticed this is causing a lint failure (`structtag: struct field BucketsMap repeats json tag "buckets" also at response.go:25 (govet)`). I was thinking of handling `buckets` as a non standard field in the switch case to work around that. Do you think that's a good approach?

Reply:
You can also make `buckets` deserialized as a `interface{}` and then just have a special method that figures out the right form it has an converts it to the canonical slice format that you do the processing on.  You can go through json again or just pull out stuff directly from the map.